File: Database/nastya/src/dbentities/Condition.java
package dbentities;

import java.io.Serializable;

/**
 * Объект, представляющий собой то самое содержание ключа, которое определяет
 * сравнение
 *
 * @author nastya
 *         Date: 30.08.2009
 *         Time: 3:20:30
 *
 * @see database.index.DatabaseKey
 */
public class Condition implements Comparable, Cloneable, Serializable{
    private String lastName;
    private String name;
    private String middleName;

    /**
     * Представляет с собой "бесконечность". То есть в случае, если это условие
     * сравнивается "снизу", толюбой элемент больше него. В случае, если "сверху",
     * то любой элемент считается больше него. Это хитрый способ не очень укладывается
     * в понятие сравниваемости
     */
    public static Condition INFINITY = new Condition("", "", "");

    public Condition(String lastName, String name, String middleName) {
        this.lastName = lastName;
        this.name = name;
        this.middleName = middleName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getMiddleName() {
        return middleName;
    }

    public void setMiddleName(String middleName) {
        this.middleName = middleName;
    }

    public int compareTo(Object o) {
        if(o == null) return 1;
        if(! (o instanceof  Condition)) return 1;
        if(this.equals(INFINITY)) return -1;
        Condition c= (Condition) o;
        if(c.equals(INFINITY)) return -1;
        int ln = lastName == null ? -1 : lastName.compareTo(c.lastName);
        int n = name== null ? -1 : name.compareTo(c.name);
        int mn = middleName== null ? -1 : middleName.compareTo(c.middleName);
        if(ln != 0) return ln;
        if(n != 0) return n;
        return mn;
    }

    public Condition clone(){
        return new Condition(lastName, name, middleName);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Condition condition = (Condition) o;

        if (lastName != null ? !lastName.equals(condition.lastName) : condition.lastName != null) return false;
        if (middleName != null ? !middleName.equals(condition.middleName) : condition.middleName != null) return false;
        if (name != null ? !name.equals(condition.name) : condition.name != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = lastName != null ? lastName.hashCode() : 0;
        result = 31 * result + (name != null ? name.hashCode() : 0);
        result = 31 * result + (middleName != null ? middleName.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Condition(" + lastName + ", " + name + ", " + middleName + ")";
    }
}

File: branches/2008-2009/Database/Babaka/DB/DB3/Entities/Condition.cs
using System;

namespace DB3.Entities
{
    [Serializable]
    public class Condition: ICloneable, IComparable
    {
        public string Name { get; set; }
        public string LastName { get; set; }
        public string MiddleName { get; set; }

        private static readonly Condition Infinity = new Condition("", "", "");

        public Condition(string name, string middleName, string lastName)
        {
            LastName = lastName;
            Name = name;
            MiddleName = middleName;
        }

        public int CompareTo(Object o) 
        {
            if(o == null) 
                return 1;
            if(! (o is Condition)) 
                return 1;
            if(this.Equals(Infinity)) 
                return -1;

            Condition c= (Condition) o;

            if(c.Equals(Infinity)) 
                return -1;

            int ln = LastName == null ? -1 : LastName.CompareTo(c.LastName);
            int n = Name == null ? -1 : Name.CompareTo(c.Name);
            int mn = MiddleName == null ? -1 : MiddleName.CompareTo(c.MiddleName);

            if(ln != 0) 
                return ln;
            if(n != 0) 
                return n;
            return mn;
        }

        //public static bool operator ==(Condition fst, Condition snd)
        //{
        //    return fst.Equals(snd);
        //}

        //public static bool operator !=(Condition fst, Condition snd)
        //{
        //    return !(fst == snd);
        //}

        //public static bool operator >(Condition fst, Condition snd)
        //{
        //    return (fst.CompareTo(snd) > 0);
        //}

        //public static bool operator <(Condition fst, Condition snd)
        //{
        //    return (fst.CompareTo(snd) < 0);
        //}

        public object Clone()
        {
            return new Condition(Name, MiddleName, LastName);
        }

        public override bool Equals(Object o) 
        {
            if (this == o) 
                return true;

            if (o == null || GetType() != o.GetType()) return false;

            Condition condition = (Condition) o;

            if (LastName != null ? !LastName.Equals(condition.LastName) : condition.LastName != null) return false;
            if (MiddleName != null ? !MiddleName.Equals(condition.MiddleName) : condition.MiddleName != null) return false;
            if (Name != null ? !Name.Equals(condition.Name) : condition.Name != null) return false;

            return true;
        }

        public override int GetHashCode()
        {
            int result = LastName != null ? LastName.GetHashCode() : 0;
            result = 31 * result + (Name != null ? Name.GetHashCode() : 0);
            result = 31 * result + (MiddleName != null ? MiddleName.GetHashCode() : 0);
            return result;
        }

        //public override string ToString()
        //{
        //    return "Condition(" + LastName + ", " + Name + ", " + MiddleName + ")";
        //}

    }
}
